feat(overlayfs): add persistent device overlay support

nadzyah opened the pull request:
This pull request adds persistent device overlay support to the overlayfs module, enabling compatibility with `cloud-initramfs-tools` overlayroot syntax for users migrating from this tool to dracut.

## Changes

- Add `overlayroot=` parameter supporting LABEL=, UUID=, PARTLABEL=, PARTUUID=, device paths, and `device:dev=` syntax
- Device-backed overlays mounted at `/run/overlayfs-backing` with symlinks for transparency
- Device resolution validation with graceful fallback to tmpfs on failure
- Comprehensive test suite covering tmpfs, LABEL, UUID, and fallback scenarios
- Documentation added to dracut.cmdline man page

## Checklist
- [x] I have tested it locally
- [x] I have reviewed and updated any documentation if relevant
- [x] I am providing new code and test(s) for it


This is my first contribution to dracut-ng, so I would appreciate any feedback and suggestions :)

FGrose commented on modules.d/70overlayfs/prepare-overlayfs.sh:
```diff
@@ -5,17 +5,89 @@ command -v getarg > /dev/null || . /lib/dracut-lib.sh
+
+        mkdir -m 0755 -p /run/overlayfs
+        mkdir -m 0755 -p /run/ovlwork
+        if [ -n "$reset_overlay" ]; then
```
The current code has the  `&& [ -h /run/overlayfs ]` condition to skip resetting a virtual overlay.

My [development version](https://github.com/FGrose/dracut-ng/blob/AO/modules.d/70overlayfs/parse-overlayfs.sh) has this logic:
```shell
[ -h /run/overlayfs ] || {
    # For temporary overlays:
    mkdir -m 0755 -p /run/overlayfs
    mkdir -m 0755 -p /run/ovlwork
}

if getargbool 0 rd.overlayfs.reset && [ -h /run/overlayfs ]; then
    ovlfsdir=$(readlink /run/overlayfs)
    info "Resetting the OverlayFS overlay directory."
    rm -r -- "${ovlfsdir:?}" > /dev/kmsg 2>&1
    mkdir -p "$ovlfsdir"
fi
```

Conan-Kudo commented:
Can we merge this into `dmsquash-live-autooverlay`?

FGrose commented on modules.d/70overlayfs/prepare-overlayfs.sh:
```diff
@@ -5,17 +5,94 @@ command -v getarg > /dev/null || . /lib/dracut-lib.sh
+            ;;
+        /dev/* | LABEL=* | UUID=* | PARTLABEL=* | PARTUUID=*)
+            overlay_mode="device"
+            overlay_device=$(label_uuid_to_dev "$overlayroot")
```
This function will echo the input (it doesn't return zero length) or return a path whether or not the partition exists.
You probably want to test `[ -b "$overlay_device ] || warn ...` before proceeding...

Same issue at line 39.

jozzsi commented:
> 1. Should `overlayroot=LABEL=persist` work alongside `rd.live.overlay=LABEL=persist` for live boot, essentially treating them as equivalent or complementary?

I think they should be equivalent if possible. Do you have any concern by making them equivalent ? 

> 2. Would the auto-partitioning/formatting logic from `dmsquash-live-autooverlay` be triggered by either parameter, or should they remain separate mechanisms?

triggered by either parameter

> 3. Is the goal to eventually phase out one of these parameters in favour of the other, or to maintain both with shared underlying logic?

I am personally ok for maintaining both with shared underlying logic for a while (e.g. 3-4 releases). The project has other aliases for command line options, but would be nice to eventually deprecate one of the syntaxes. 

See https://github.com/dracut-ng/dracut-ng/blob/main/man/dracut.cmdline.7.adoc?plain=1#L1456 for deprecated aliases

jozzsi reviewed (approved):
non-owner LGTM.

Thanks @nadzyah for working with us on this project, this is an important PR.

bdrung commented on man/dracut.cmdline.7.adoc:
```diff
@@ -1109,6 +1109,44 @@ The **rd.overlayfs.readonly** option, which allows a persistent overlayfs to
 be mounted read-only through a higher level transient overlay directory, has
 been implemented through the multiple lower layers feature of OverlayFS.
 
+**overlayroot=**__{tmpfs|LABEL=<label>|UUID=<uuid>|PARTUUID=<uuid>|PARTLABEL=<label>|/dev/<device>|device:dev=<device>|disabled}__::
```
To clarify my idea: Move `rd.live.overlay` out of the `rd.live` namespace and pick something in the `rd.` namespace (`rd.overlay` or `rd.overlayfs.root` or ???). For compatibility with `cloud-initramfs-tools` `overlayroot` can be added as alias to that variable in the `rd.` namespace.

bdrung commented on test/TEST-21-OVERLAYFS/test.sh:
```diff
@@ -0,0 +1,76 @@
+        -initrd "$TESTDIR"/initramfs.testing
+
+    test_marker_check
+    test_marker_reset
```
In addition to the test marker check, can you check that the overlay was mounted correctly? E.g. write something on the overlay and afterwards check the overlay.img?

bdrung commented on test/TEST-21-OVERLAYFS/test-overlayfs.sh:
```diff
@@ -0,0 +1,90 @@
+#!/bin/sh
```
This new test 21 should be added to the CI in .github/workflows/.

bdrung commented on .github/workflows/integration.yml:
```diff
@@ -91,6 +91,7 @@ jobs:
                     - "12"
                     - "13"
                     - "20"
+                    - "21"
```
Can you also add this test to daily-basic (basic+extended), daily-hostonlystrict, and daily-omitsystemd?

bdrung commented on test/TEST-21-OVERLAYFS/test.sh:
```diff
@@ -0,0 +1,50 @@
+    client_run "tmpfs overlay" "rd.overlayfs=1"
+    client_run "persistent device overlay (LABEL)" "rd.overlay=LABEL=OVERLAY"
+    client_run "persistent device overlay (UUID)" "rd.overlay=UUID=$OVERLAY_UUID"
+    client_run "fallback to tmpfs (non-existent LABEL)" "rd.overlay=LABEL=NONEXISTENT"
```
You could also test the device path overlay:
```
    client_run "persistent device overlay (device path)" \
        "rd.overlay=/dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_overlay"
```

bdrung reviewed (changes requested):
Thanks. Overall this change looks good to me. I have some smaller change requests and a few open questions.

bdrung commented on modules.d/70overlayfs/prepare-overlayfs.sh:
```diff
@@ -5,17 +5,77 @@ command -v getarg > /dev/null || . /lib/dracut-lib.sh
+overlay_mode="tmpfs"
+overlay_device=""
+
+if [ -n "$overlay" ]; then
```
If `overlay` is not set, you could exit early. Then you do not have to set `overlayfs="yes"`.

nadzyah committed:
feat(overlayfs): add persistent device overlay support

Extends the overlayfs module to support persistent overlay storage on
block devices using rd.overlay=. Supported formats: LABEL=, UUID=,
PARTUUID=, PARTLABEL=, and /dev/ paths.

Changes written to the overlay device persist across reboots. Falls
back to tmpfs if the device cannot be resolved or mounted.

bdrung reviewed (dismissed):
Looks good to me now. Thanks for contributing and writing tests. 🥳

nadzyah commented:
@bdrung I've rebased the changes since the #2153 PR got merged
